commit 7ea26f035e6dce017e00f743a29a79c27940fdce (refs/changes/63/1163/3)
Author: Alex Wilson <alex.wilson@joyent.com>
Date:   2016-12-22T11:20:40-08:00 (2 years, 10 months ago)
    
    joyent/node-cueball#47 rapid add/remove of backend in DNS leads to ConnectionSet crash
    Reviewed by: Bryan Cantrill <bryan@joyent.com>

diff --git a/lib/resolver.js b/lib/resolver.js
index cf15e6e..840d67a 100644
--- a/lib/resolver.js
+++ b/lib/resolver.js
@@ -877,6 +877,10 @@ CueBallDNSResolver.prototype.state_process = function (S) {
 	}
 
 	this.emit('updated');
+
+	/* Write down what we did to help debugging. */
+	this.r_lastProcessed = { added: added, removed: removed };
+
 	S.gotoState('sleep');
 };
 
diff --git a/lib/set.js b/lib/set.js
index daf6366..cb74cf9 100644
--- a/lib/set.js
+++ b/lib/set.js
@@ -141,9 +141,15 @@ CueBallConnectionSet.prototype.on_resolver_removed = function (k) {
 	});
 	delete (this.cs_dead[k]);
 	var fsms = self.cs_fsms[k];
-	mod_assert.ok(fsms.length <= 1);
-	if (cks.length === 0 && fsms[0] && !fsms[0].isInState('idle'))
-		fsms[0].close();
+	/*
+	 * We might not have anything in cs_fsms for this backend (it's all
+	 * populated async and only if the planner says we should).
+	 */
+	if (fsms !== undefined) {
+		mod_assert.ok(fsms.length <= 1);
+		if (cks.length === 0 && fsms[0] && !fsms[0].isInState('idle'))
+			fsms[0].close();
+	}
 };
 
 CueBallConnectionSet.prototype.isDeclaredDead = function (backend) {
diff --git a/package.json b/package.json
index e14bf6d..0df5f04 100644
--- a/package.json
+++ b/package.json
@@ -1,6 +1,6 @@
 {
   "name": "cueball",
-  "version": "1.2.2",
+  "version": "1.2.3",
   "description": "",
   "main": "lib/index.js",
   "dependencies": {
